contributors/design-proposals/network/service_port_ranges.md

```diff
@@ -0,0 +1,128 @@
+  protocol: UDP
+  ...
+- name: sip-udp5559
+  containerPort: 5559
+  protocol: UDP
+```
+
+Doing above for 500 ports is not pretty. Can we have a way to allow port ranges like 5060-5559?
+
+## Objective
+
+This proposal builds off of earlier requests to [Support port ranges in services](https://github.com/kubernetes/kubernetes/issues/23864) and aims to document all the requirements of supporting port ranges in services. Each port range is a single port or single contiguous port range, ranging from low to high, for which this forwarding rule matches. Packets of the specified protocol sent to these ports will be forwarded on to the appropriate target instances.
+
+## Goals
+
+A server side solution to support single or multiple contiguous port range.
+
+## Non-goal
+
+Target port remapping in ranges.
+
+Pod port ranges.
+
+## User Cases
+
+* Specifying port ranges in services declarations is mandatory for any VoIP-related application.
+* With the standard RTP port range which is 16384-32767. Clients connect on the standard RTP port range with a media server application.
+* Needs this change to make TeleStax (an RTC framework) work well in Kubernetes.
+
+## Proposal
+
+This proposal builds off earlier feature requrest to [Support port ranges in services](https://github.com/kubernetes/kubernetes/issues/23864). We aims to implement service port ranges at least for iptables and IPVS modes.
+
+### Service API changes
+
+Create a new filed `portRangeSize` in `ServicePort`.
+
+
+```go
+type ServiceSpec struct {
+  Ports []ServicePort
+}
+
+type ServicePort struct {
+  // portRangeSize is the contiguous ports number that are exposed by this service. Given port = 80 and portRangeSize = 100, it means ports 80-179, 100 ports in total, will be exposed. In this case, port means the starting port of a range.
+  // portRangeSize is default to be 1 for compatibility.
+  // If nodePort is specified, then nodePort is the starting port of range and nodePort+portRangeSize-1 is the ending port of range.
+  // A service can expose multiple port ranges and each port range should not be overlapped.
+  // Service port range does not support target port remapping, it implies that targetPort should not be specified or should be equal to service port.
+  // +optional
+  PortRangeSize int `json: "portRangeSize,omitempty"`
+}
+```
+
+### Service port ranges example
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: service-port-ranges
+  namespace: default
+spec:
+  ports:
+  - name: range1
+    port: 90
+    protocol: TCP
+    portRangeSize: 100
+    protocol: UDP
+  - name: range2
+    port: 200
+    portRange: 300
+    protocol: TCP
+```
+
+### Kube-proxy implementation limitations
+
+NodePort and ExternalIPs Services don't support port ranges since kube-proxy will hold local ports open for these services(nodeport and "external" IP happens to be an IP that is local to this machine). It's challenging to keep some huge port ranges open on host.
```

Comment:
The SIP/RTP case that people talk about a lot requires having some sort of externally-reachable IP address. So, if we have this restriction, is the feature still useful?

(I think we need much more detailed and specific use cases.)

Reply:
Can you link to the mention of this external IP mention? If this is the case, does that mean this is more of a feature request to ingress?

Reply:
I think this is a major limitation of the "concentric" model for NodePorts and LoadBalancer services.  I don't see a simple resolution except to allow those platforms that don't need the concentricness to opt out of them.

e.g. Google cloud doesn't need NodePort for LoadBalancer, but AWS does (or did).  I could add an annotation to say "cloud.google.com/bypass-node-port: true".  Or we could allow `nodePort: -1` to mean "skip it for this port".

I think that this needs to be built into the next generation of Service API.

Reply:
> Can you link to the mention of this external IP mention? If this is the case, does that mean this is more of a feature request to ingress?

Sorry, missed this question earlier. Eg, the initial comment on this PR refers to #23864, but the person who filed that bug [said](https://github.com/kubernetes/kubernetes/issues/23864#issuecomment-206109433) "I do need to target the ports through a Service since that is the only way outside users would be able to place a call via SIP". So a solution that doesn't allow reserving the ports on some public IP isn't going to help them.